### ISSUE TITLE
delay() not functioning as expected.

### ISSUE THREAD

**Issue by dschalk** [cujojs/most#200]
I have been having trouble with postponing changes to the dom in my motorcycle.js app running at http://schalk.net:3099, repo at https://github.com/dschalk/JS-monads-part4. I haven't been able to get functions that rely on setTimeout to work at all, or to update the DOM when they are supposed to.  I'll try promises next.

I am raising an issue here because, when I tried `most.delay()', every instance of its use was triggered at the time specified in the first instance. Here is the code:
```javascript 
 const mult2$ = mMmult.x.result.map(v => {
    mMmult.x.product2 = ret(v).x
  });

  const mult3$ = mMmult.x.result.map(v => {
    mMmult.x.product2 = ret(v)
    .bnd(add, 1000)
  }).delay(2000);

  const mult4$ = mMmult.x.result.map(v => {
    mMmult.x.product2 = ret(v)
    .bnd(add, 1000)
    .bnd(double).x
  }).delay(10000);

  const mult5$ = mMmult.x.result.map(v => {
    mMmult.x.product2 = ret(v)
    .bnd(add, 1000)
    .bnd(double)
    .bnd(add, 1).x
  }).delay(15000);
```
The first result displays correctly. Then, after two seconds, the result from the last one (delay(15000)) displays. It is as though the first time interval specified is applied in all subsequent instances.

**Comment by campersau** [comment]
I think you have to `delay` first and then `map`:

```js
const mult2$ = mMmult.x.result.map(v => {
    mMmult.x.product2 = ret(v).x
  });

  const mult3$ = mMmult.x.result.delay(2000).map(v => {
    mMmult.x.product2 = ret(v)
    .bnd(add, 1000)
  });

  const mult4$ = mMmult.x.result.delay(10000).map(v => {
    mMmult.x.product2 = ret(v)
    .bnd(add, 1000)
    .bnd(double).x
  });

  const mult5$ = mMmult.x.result.delay(15000).map(v => {
    mMmult.x.product2 = ret(v)
    .bnd(add, 1000)
    .bnd(double)
    .bnd(add, 1).x
  });
```

As the example in the [README](https://github.com/cujojs/most/blob/master/README.md) shows:

```js
// After 4 seconds, logs 10
most.from([1, 2, 3, 4])
    .delay(1000)
    .reduce(function(result, y) {
        return result + y;
    }, 0)
    .then(function(result) {
        console.log(result);
    });
```

**Comment by dschalk** [comment]
That looks better. I think I tried doing it the right way and still didn't get a good result. Lots of late night, frustrating experimentation has receded into a distant fog. Algorithms that worked in my plain Snabbdom application all flopped.

I opted for doing things more in the Cycle/Motorcycle way. See http://schalk.net:3099, repo at https://github.com/dschalk/JS-monads-part4.  The relevant code is:
```javascript
  const unitDriver = function () {
    return periodic(1000, 1);  // Creates a stream of 1-1-1-1..., in one-second intervals.
  }
  
  const sources = {
    DOM: makeDOMDriver('#main-container'),
    WS: websocketsDriver,
    UNIT: unitDriver           // Added unitDriver to the sources object.
  }

  const unitAction$ = sources.UNIT.map(v => {  // unitDriver, cycled back around from "run".
      mMunit.ret(mMunit.x + v)  // mMunit is a monad dedicated to calling "next" with its bnd method.
      .bnd(next, 2, mMZ26)  // Releases mMZ26 (below) after two seconds.
      .bnd(next, 4, mMZ27)
      .bnd(next, 6, mMZ28)
      console.log('mMunit.x ', mMunit.x)
  })

  const mult2$ = mMmult.x.result.map(v => {
    mMmult.x.product2 = v;
    let mMtemp = new Monad(v);
    mMZ26.bnd(() => mMtemp.bnd(add, 1000).bnd(mMtemp.ret).bnd(x => mMmult.x.product2 = x));
    mMZ27.bnd(() => mMtemp.bnd(double).bnd(mMtemp.ret).bnd(x => mMmult.x.product2 = x));
    mMZ28.bnd(() => mMtemp.bnd(add, 1).bnd(mMtemp.ret).bnd(x => mMmult.x.product2 = x)); 
    mMunit.ret(0);
  })
```
I put this comment in [JS-monads-part4](http://schalk.net:3099): "Algorithms that worked in JS-monads-part3, a plain Snabbdom application, don't work in Motorcycle.js. But Motorcycle.js is a whirling wonder, awe-inspiring to behold and to use."

**Comment by nissoh** [comment]
Seems like you forgot to flatten your inner stream(assuming map is related to most's operator).
you can do that by using the `switch` operator

````
  const mult3$ = mMmult.x.result.map(v => {
    mMmult.x.product2 = ret(v)
    .bnd(add, 1000).delay(2000)
  }).switch();
````

BTW.. you might wanna look at the new `::` operator
https://github.com/zenparsing/es-function-bind

**Comment by dschalk** [comment]
This also works:
```javascript
  const mult3$ = mMmult.x.result.delay(2000).map(v => {
    mM27.ret(mMmult.x.product)
  })  
  
  const mult4$ = mMmult.x.result.delay(4000).map(v => {
    mM27.bnd(add, 1000).bnd(mM27.ret)
  })  
  
  const mult5$ = mMmult.x.result.delay(6000).map(v => {
    mM27.bnd(add, 1000).bnd(double).bnd(mM27.ret)
  })  
  
  const mult6$ = mMmult.x.result.delay(8000).map(v => {
    mM27.bnd(add, 1).bnd(mM27.ret)
  })  
```
I didn't need to create a driver. "delay" works as expected in Motorcycle.js.

**Comment by dschalk** [comment]
I experimented some more with "dalay". Streams did not resume until after the specified milliseconds had elapsed, but then I could not depend on computation sequences to run only once. It is as though delay(n) means "continue once, or twice, or maybe more times after n milliseconds have elapsed." I have also been unable to get reliable results, or any results at all, using setInterval or setTimeout. 

I am very pleased with Motorcycle, and will continue to use it. My impression at this point is that my slolution (above) is not too convoluted. Defining a driver like my unitDriver and an action stream that receives values from it is the only way to get dependable results, as far as I can tell. 

My solution didn't require very many lines of code, and the driver can be re-used any time I want to delay the appearance of something in the DOM. But if there is a simpler way to reliably postpone future actions, I would much appreciate it if somebody would clue me in.

**Comment by dschalk** [comment]
This isn't the best place for this discussion, but I'll add what is probably my last comment here. I just want to show how the result is calculated in http://schalk.net:3099 without resort to MonadIter. Here it is:
```javascript
  const mult3$ = mMmult.x.result.map(v => {
    mMtem.ret(v)
    mMmult.x.product3 = v;
    mMpause.ret(0);
  })

  const mult4$ = sources.UNIT.map(v => {
      mMpause.ret(mMpause.x + v)
      if(mMpause.x ===1) {
        mMtem.bnd(add, 1000).bnd(mMtem.ret).bnd(x => mMmult.x.product3 = x)
      }
      if(mMpause.x === 2) {
        mMtem.bnd(double).bnd(mMtem.ret).bnd(x => mMmult.x.product3 = x)
      }
      if(mMpause.x === 3) {
        mMtem.bnd(add, 1).bnd(x => mMmult.x.product3 = x) 
      }
    })
```
This uses "unitDriver", defined above.

**Comment by briancavalier** [comment]
Hey @dschalk.  Appreciate all the examples in this thread.  Is it possible you can provide a simplified example that reproduces the problem(s) you've described?  For example, you could use [requirebin](http://requirebin.com) or something that could be pasted directly into a js file and run with node locally.

I'm quite confident in delay's behavior (the [circles example](https://github.com/cujojs/most/tree/master/examples/circles) is a seriously intense stress test of it), but I'd like to see if there may be an edge case we've missed.  It could also be that your expectations about what delay is supposed to do don't match what it actually does.  That could mean that we need to improve our documentation.

Anyway, it'd be really helpful to have a simple, runnable test case.

I've noticed that in your examples, you seem to be using shared mutable state and side effects.  For example, the purpose of `map` is to turn events of type A into events of type B using a pure function (e.g. it should only use its input--the event--and should return something), rather than to mutate shared properties (e.g. `mMult.x.product3 = x`).

Using side effects in a highly asynchronous system can be very difficult to reason about.  One goal of reactive programming is to forego side effects, in favor of purely transformative data flow style.  That made me wonder if some of the oddities you are experience are due to using a side-effectful approach.

I hope that didn't come across as critical, but rather as just trying to understand the situation, with the goal of trying to help!

**Comment by dschalk** [comment]
The erratic behavior occurs only when I make a change that starts setTimeout or delay before a previous setTimeout() or delay has completed. window.clearTimeout() is ignored in my up-to-date versions of Chrome and Firefox. When I change a number in http://schalk.net:3099 using BACKSPACE and then entering a substitute number, a computation starts immediately after the backspce key is pressed and then again when the new number is entered. The base number gets 1000 added to it twice and gets doubled twice. This code works perfectly if I wait for completion after pressing the backspace key:
```javascript
  const mult5$ = mMmult.x.result
  .filter(x => x !== 0).map(v => {mM27.ret(v)}).delay(1000)
  .map(() => mM27.bnd(add, 1000).bnd(mM27.ret)).delay(1000)
  .map(() => mM27.bnd(double).bnd(mM27.ret)).delay(1000)
  .map(() => mM27.bnd(add, 1).bnd(mM27.ret)).delay(1000)
```
mM27 is an instance of Monad. Using its "bnd" method with add and double do what you would expect, but they return an anonymous monad holding the result in its "x" attribute. "bnd(mM27.ret)" uses mM27's "ret" method to fetch the result from the anonymous monad.

clearTimeout's not working in Motorcycle doesn't seem to be relevant to 'most', although it might make it hard to fashion an "abort delay" function that works in Motorcycle. I think an "abort delay" function might be a useful addition to `most`. My solution using a stream of 1's to cause delays is interesting, but does seem excessively elaborate.

For a while, my monad `ret` method created a new monad with the same id as the calling monad. Browsers don't let you use `caller`, so I was using `eval` with the id, or window[id], to give the new monad the id of the monad using its `ret` method. I could have kept the monad and substituted a new value for the old one, rather than simply mutating the old one. I don't mutate the monads, but I went back to mutating their values. I don't see what is lost by doing this. In my server, I keep state in a TMVar. The TMVar doesn't mutate, even though its value frequently changes. Javascript allows mutation of object attributes, which strikes me as being a convenient shortcut. I am pretty sure that I am confessing my ignorance, and not presenting a sound argument. I wish I had an example of mutating an attribute's causing a problem in Cycle.js or Motorcycle.js. 

My problem stems from the fact that I use `delay` in a sequence of computations, and I can't abort an ongoing sequence when a new one starts. That causes some computation steps to occur twice. I can get the result I want, but not by using setTimeout or `delay`.  

You certainly didn't come across as being unfairly critical. Would the immutable library that people use with React be useful in a Motorcycle application?

**Comment by dschalk** [comment]
The code below is now part of the app running at http://schalk.net:3099 . It consistently gives the desired results because `debounce` provides enough time for any ongoing computations to complete.
```javascript
  const mult5$ = mMmult.x.result
  .debounce(3200).map(v => {mM27.ret(v)}).delay(1000)
  .map(() => mM27.bnd(add, 1000).bnd(mM27.ret)).delay(1000)
  .map(() => mM27.bnd(double).bnd(mM27.ret)).delay(1000)
  .map(() => mM27.bnd(add, 1).bnd(mM27.ret)).delay(1000)
```
I'm starting work on JS-monads-part5. It will feature immutable data structures using the Facebook immutable-js library.

**Comment by briancavalier** [comment]
@dschalk [v0.18.2](https://github.com/cujojs/most/releases/tag/0.18.2) just landed with a fix for a very subtle issue related to `delay`.  The issue seems like it'd be rare, but it's possible you were encountering it.  Maybe give the update a try and see what happens.

> Would the immutable library that people use with React be useful in a Motorcycle application?

Sure.  Immutable.js is a general purpose immutable data structures library, so it's quite useful in almost any context :)

**Comment by dschalk** [comment]
I upgraded to v0.18.2. Nothing changed. If I don't wait for an already-started computation sequence to finish, I get larger than desired results due to both computation sequences calling `add` and `double`.

I don't know if it is feasible, but a "clearWait" function at the beginning of my computation sequence would make wait more useful in cases such as this.

**Comment by briancavalier** [comment]
@dschalk It was a rare situation, so I'm not surprised nothing changed.

It's still a bit tough to understand what you're seeing without a runnable example.  Any chance you could put one together?

**Comment by dschalk** [comment]
@briancavalier  I changed the definition of Monad's `ret` method and added a demonstration to http://schalk.net:3099 showing that you can move back through history by selecting previous versions of a monad instance which is saved in an array every time it calls its `ret` method. Monad might not be familiar to you, but it is very simple. I used its `ret` method in a demonstration I made for you.

It is all set forth at the bottom of the page at http://schalk.net:3099.

**Comment by dschalk** [comment]
Update: Nothing is being mutated in the app running online at http://schalk.net:3099. 

I found an aberation in this:
```javascript
  const mult5$ = mMmult.x.result
  .map(v => {mM27.ret(v)})
  .map(() => mM27.bnd(add, 1000).bnd(mM27.ret)).debounce(1000)
  .map(() => mM27.bnd(double).bnd(mM27.ret)).debounce(1000)
  .map(() => mM27.bnd(add, 1).bnd(mM27.ret)).debounce(2000)```

**Comment by briancavalier** [comment]
@dschalk Thanks, I appreciate the effort you're putting into your app and into learning algebraic structures.  FWIW, most.js Streams are a (or more properly, "have a") Functor, Applicative Functor, Monoid, and Monad per the [fantasy-land specification for JS algebraic structures](https://github.com/fantasyland/fantasy-land)

There's quite a bit there on your app's page.  That's cool.  However, it also makes debugging extremely difficult.  I think our goal needs to be to find a way to create a smallest possible case that reproduces a problem.  Without doing that, it seems impossible determine whether most, motorcycle, or your own code is the source of the issue.

As a next step, can you take the last debounce example that you mentioned and try to isolate it?  For example, are you able to reproduce the same issue if you remove motorcycle from the picture and implement a similar example using on most and `console` (instead of DOM manipulation), etc.

**Comment by dschalk** [comment]
@briancavalier  I just now eliminated some mutating code in the history example. I don't think any mutating remains now.

I am not able to reproduce the `debounce` issue now, with motorcycle in the picture. While rapidly changing the numbers, I estimate at least 100 times, three times the result was off a little. I know it was 2 greater than expected twice and maybe all three times. I don't know what I did to produce the anomaly, and I don't know that doing exactly the same thing again would produce the unexpected result again. I suspect a previously initiated sequence performed a computation. That certainly is what happens in the example at the bottom of the page using `delay`. It could very well be that motorcycle was the cause in the debounce example.

I will gladly create a debounce example that prints to the browser console, or the system console if you prefer. I don't understand what you mean by "using on most". Do you mean "using most"?

I am looking forward to studying the most code. I am somewhat familiar with Haskell Functor, Applicative Functor, Monoid, and Monad. My http://schalk.net:3099 server is built on the  [wai-websockets](https://github.com/yesodweb/wai/tree/master/wai-websockets)

**Comment by dschalk** [comment]
I tried some more, but the algorithm using `debounce` consistently gives the expected result. I wonder if the rare exceptions occurred when a new sequence started on the same tick when a live debounce waiting period expired, resulting in a previously scheduled computation being performed on the freshly generated number. Or something similar to that.
